partiql-planner/README.adoc

```diff
@@ -0,0 +1,76 @@
+= Metadata
+
+The _metadata_ package contains the interfaces for PartiQL's system
+ which enables developers to leverage the planner across diverse backends.
+
+== Classes
+
+[cols="1e,4"]
+|===
+| Class | Description
+
+| Scope | A _Scope_ is a namespace for tables and routines. It may contain nested scopes, and a top-level scope is called a _Catalog_. The initial _Scope_ interface provides methods for table, function, and sub-scope lookup.
+| Table | A _Table_ can be either a base table, view, or index. Every table has a _Shape_ and a _Statistic_.
+| Function | A _Function_ is a PartiQL-routine which can be invoked from an expression context. A _Function_ is unique within a scope by its name and formal arguments (parameters); therefore its name may be overloaded like in SQL.
+| Statistics | A _Statistics_ object is used by the query planner for table statistics and metadata such as available keys and indexes. This is currently a placeholder.
+| Session | The _Session_ holds the current scope and PATH for name resolution.
+
+|===
+
+== Name Resolution
+
+The following section gives a high-level description of the name resolution rules for tables and routines.
+
+NOTE: A name is composed of _n_ steps and can be qualified (_n > 1_) or unqualified (_n = 1_). A _relative_ name begins with a step whereas an _absolute_ name begins with a leading period.
+
+[source]
+----
+name
+ :     id ('.' id*)*  # relative
+ | '.' id ('.' id*)*  # absolute
+ ;
+----
+
+[cols="1e,4"]
+|===
+| Type | Example
+
+| Relative Name |  `n~1~.---.n~n~`
+| Absolute Name | `.n~1~.---.n~n~`
+
+|===
+
+=== Table Resolution
+
+Let _N_ be a name with steps _n~1~ . n~2~ ... n~i~_ and let _S_ be the session scope with name _s~1~_ ... _s~j~_.
+
+. Let _T_ be the fully-qualified table name.
+.. If _N_ is a relative name, then _T = s~1~ . s~2~ ... s~j~ . n~1~ . n~2~. ... n~i~_ .
+.. If _N_ is an absolute name, then _T = n~1~ . n~2~ ... n~i~_ .
+. Let _n_ be the number of steps in _T = t~1~ . t~2~ ... t~n~_ .
+. Find the largest _m ≤ n_ such that _t~1~ . t~2~ ... t~m~_ is a database name.
+. Any remaining _t~m+1~ ...t~n~_ are path navigations.
+
+NOTE: Examples to follow.
+
+=== Routine Lookup
+
+Let _N_ be a routine name with steps _n~1~ . n~2~ ... n~i~_ and let _S_ be the session scope with name _s~1~_ ... _s~j~_.
+
+. If n = 1 then the session path is searched left-to-right for _n~1~_.
+. If n > 1, then let _R_ be the fully-qualified routine name.
+.. If _N_ is a relative name, then _R = s~1~ . s~2~ ... s~j~ . n~1~ . n~2~. ... n~i~_ .
+.. If _N_ is an absolute name, then _R = n~1~ . n~2~ ... n~i~_ .
+. Lookup the fully-qualified name _R_ and fail compilation if not found.
```

Comment:
Regarding "fully-qualified" vs "absolute", I think the following suggestion is a bit more clear.

```suggestion
Let _N_ be the routine invocation name (absolute or relative) with steps _n1_, _n2_, …​, _ni_. Let _S_ be the session scope with name _.s1.s2. …​ sj_. Let _R_ by the absolute name of the invoked routine.

1. If _N_ is an absolute name, then _R = . n1 . n2 …​ ni_ . Return _R_.
2. If n = 1 then the function path is searched left-to-right for _n1_.
3. If n > 1, then _R = . s1 . s2 …​ sj . n1 . n2. …​ ni_ .
4. Lookup the absolute name _R_. Return _R_ if found. Fail compilation if not found.
```

Maybe the "function path" should be SQL Path or invocation path. Up to you.

Reply:
Typo, it should have been "session path"